### DISCUSSION TITLE
No devices discovered on Windows

### DISCUSSION THREAD

**Discussion by 0ge** [hbldh/bleak#463] [Q&A]
* bleak version: 0.9.0, 0.9.1, 0.10.0
* Python version: 3.9
* Operating System: Win10

### Description

I'm running the [scanning example](https://bleak.readthedocs.io/en/latest/scanning.html) on a Windows machine.
However, it does not discover any devices.

### What I Did

Installed bleak with `pip install bleak`.
Running the example from both VSCode and command prompt. (Different command to enable logging)

I've tried with `$env:BLEAK_LOGGING=1` and also `logging.basicConfig(level=logging.DEBUG)` but I don't see any logging.
The only output is `DEBUG:asyncio:Using proactor: IocpProactor`

It runs on another machine, so I'm thinking there is something related to the installation?
Unfortunately, I did not do the installation on the other machine.

Any help would be greatly appreciated.

**Comment by dlech** [comment]
Maybe Bluetooth is turned off on that machine?

**Reply by 0ge** [reply]
No, its turned on.
(Thanks for letting me know of the discussions feature of GitHub!)

**Reply by 0ge** [reply]
If I go to Windows Settings I can scan and discover some Bluetooth devices.

**Comment by dlech** [comment]
>     * Python version: 3.9

Try Python 3.8.

**Reply by 0ge** [reply]
I tried with 3.7.7 (didn't have 3.8). Now there is output, but still doesn't not discover any devices:
```
DEBUG:asyncio:Using selector: SelectSelector
2021-02-24 17:46:08,052 bleak.backends.dotnet.scanner DEBUG: 0 devices found. Watcher status: 3.
DEBUG:bleak.backends.dotnet.scanner:0 devices found. Watcher status: 3.
2021-02-24 17:46:08,052 bleak.backends.dotnet.scanner DEBUG: Could not remove event handlers: unknown event handler...
DEBUG:bleak.backends.dotnet.scanner:Could not remove event handlers: unknown event handler...
```

**Reply by hbldh** [reply]
How did you get it to even run on 3.9 in Windows? Did you build the `pythonnet` dependency yourself?

What Windows 10 version do you have? (`platform.win32_ver()`)

**Reply by 0ge** [reply]
The steps were something like this:
1. Download VS Code
2. Run `<path>/Python39.exe -m pip install bleak`
3. It automatically started to build `pythonnet` but failed, because I was missing MSBuild 2015 and .NET 3.5 (or something .NET related)
4. Once built, it was successful

Not sure how `pythonnet` was installed for 3.7 as I think I did this after 'fixing' 3.9, so there wouldn't be the same build issues.

```
>>> import platform
>>> platform.win32_ver()
('10', '10.0.18362', 'SP0', 'Multiprocessor Free')
```

**Reply by dlech** [reply]
From what I read, 3.9 support is there in `pythonnet` but there were some reported crashes, so they didn't publish a binary wheel.

**Reply by hbldh** [reply]
@dlech Ah, I had not seen that. Good to know.

@0ge Could you install a different branch on either 3.9 or 3.7 and run your program again:
```
pip install --force-reinstall https://github.com/hbldh/bleak/archive/feature/winrt2.zip#egg=bleak
```
This is a different way of interacting with the Windows UWP BLE. See if that makes any changes. If not, I would suspect that the problem is in Windows and not in Bleak...

**Reply by 0ge** [reply]
Still nothing. (Both 3.7 and 3.9)

**Reply by 0ge** [reply]
No ideas on what is "wrong" in Windows?

**Reply by hbldh** [reply]
Get the Bluetooth LE Explorer from the Windows Store and try that (code here: https://github.com/Microsoft/BluetoothLEExplorer). That uses the same APIs that Bleak does. If that works and Bleak doesn't, then something is most probably wrong in Bleak.
